pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/GenericToken.java

```diff
@@ -4,6 +4,33 @@
 
 package net.sourceforge.pmd.lang.ast;
 
-public class GenericToken {
+/**
+ *  Represents a language-independent token such as constants, values language reserved keywords, or comments.
+ */
+public interface GenericToken {
+
+    /**
+     * Obtain the next generic token according to the input stream which generated the instance of this token.
+     * @return the next generic token if it exists; null if it does not exist
+     */
+    GenericToken getNextGenericToken();
+
+    /**
+     * Obtain a special generic token which, according to the input stream which generated the instance of this token,
+     * precedes this instance token and succeeds the previous generic token (if there is any).
+     * @return the special token if it exists; null if it does not exist
+     */
+    GenericToken getPreviousSpecialGenericToken();
```

Comment:
the concept of special token is a little too JavaCC-related... Antlr just sends the tokens to a different channel.

Maybe we should be a little more semantic here on what we expect from grammars for PMD. Maybe something such as `getPrevComment()` or something of the sort?

Of course the docs would need to be updated

Reply:
Ok so in the cases of auto-fixes or CPD we want to preserve comments so what you say fits in perfectly. 

However, if we go through that way and the case is presented where a non-comment and "non-regular" token is needed to be used, the solution would be either to add a specific method for that case or to deprecate this method and make a more general one.

I currently cannot foresee if PMD will ever need to address this cases, so I'm just speculating.

Reply:
A special token (JavaCC) is a token that:

1. can appear at any point in the grammar
2. is not part of the AST itself

I can think of no thing other than comments that may fit this criteria under a programming language perspective. Remember, what we define here is jut what we expect from PMD grammars. I feel comfortable with such semantics.

Reply:
Perfect, I have just updated the interface's method. I will now follow to update the PR's comment